addons/project/models/project_project.py

```diff
@@ -462,13 +461,26 @@ def create(self, vals_list):
             for vals in vals_list:
                 if 'label_tasks' in vals and not vals['label_tasks']:
                     vals['label_tasks'] = task_label
-        if len(self.env.companies) > 1 and self.env.user.has_group('project.group_project_stages'):
-            # Select the stage whether the default_stage_id field is set in context (quick create) or if it is not (normal create)
-            stage = self.env['project.project.stage'].browse(self._context['default_stage_id']) if 'default_stage_id' in self._context else self._default_stage_id()
-            # The project's company_id must be the same as the stage's company_id
-            if stage.company_id:
+        if self.env.user.has_group('project.group_project_stages'):
+            if 'default_stage_id' in self._context:
+                stage = self.env['project.project.stage'].browse(self._context['default_stage_id'])
+                # The project's company_id must be the same as the stage's company_id
+                if stage.company_id:
+                    for vals in vals_list:
+                        vals['company_id'] = stage.company_id.id
+            else:
+                # If a company is set on the project, use a stage with the same company or no company
+                companies_ids = set(vals.get('company_id', False) for vals in vals_list)
+                default_stages_by_company = {company_id : self.env['project.project.stage'].search([('company_id', 'in', [False, company_id])], limit=1) for company_id in companies_ids}
+                default_stage = self._default_stage_id()
```

Comment:
we will do `self._default_stage_id()` without need if there is no False company_ids, and if there is a False we will do:

```py
self.env['project.project.stage'].search([('company_id', 'in', [False, False])], limit=1)
```
and never use it

this would complicate the dict comprehension, but you could fix both case with something like this:

```suggestion
                default_stages_by_company = {
                    company_id : self.env['project.project.stage'].search(
                        [('company_id', 'in', [False, company_id])], 
                    limit=1) if company_id else self._default_stage_id()
                    for company_id in companies_ids
                 }

Reply:
that's better, thanks :+1: